### DISCUSSION TITLE
LiveTiming in Python 3.10

### DISCUSSION THREAD

**Discussion by foxyskippy** [theOehrly/Fast-F1#245] [Q&A] [answered]
Is there a plan to fix the issues with LiveTiming in Python 3.10?

**Comment by theOehrly** [comment, marked as answer]
~~No, I don't plan to fix it, to be honest.~~ 
It will be fixed.

The problem here is not really FastF1 but signalr-client-aio which FastF1 depends on for connecting to the data live stream.
That module is incompatible wity Python 3.10 and seemingly not maintained anymore. I don't know of any alternatives. 
~~Therefore, one option is for me to fork, update and maintain that module myself. That'd require me to really dig into the SignalR spec to understand the protocol so that I even have an idea what I'm doing.
An alternative might be to switch over to some kind of client written in c/c++ and write a wrapper to access it from Python.~~

~~Both of these options are quite time consuming. They're really not worth the effort when we still have Python 3.8/3.9 working just fine.~~

Scratch this, it's not that complicated. The solution is discussed below and will likely be implemented soon.

**Reply by oscr** [reply]
I agree that there are more important tasks to focus on for the time being. That said, I had a boring meeting and attempted to upgrade it and it seems to work: https://github.com/oscr/python-signalr-client/commit/3a62e3ecc6165431155cbdd436afd475ab78bc45

Given that it actually works (I may need some help verifying) I suggest making a module out of it. Stripping everything that isn't needed making a minimal version. Once in place it would be possible to make changes, but not required.

**Reply by theOehrly** [reply]
@oscr Imo, that change makes it backwards incompatible with 3.8/3.9?

**Reply by oscr** [reply]
@theOehrly Yes, with a minor update it now runs 3.8, 3.9 and 3.10. I tested by just changing the Dockerfile version.

https://github.com/oscr/python-signalr-client/commit/45edbc4d1004372f92f5dec590abc076b63e6853

```
docker build -t signalr . && docker run signalr
(.. output excluded ..)
{'R': {'Heartbeat': {'Utc': '2022-08-28T15:18:14.5484706Z', '_kf': True}, 'CarData.z': '7ZSxDoIwEIbf5WYgd9fCla7EN9BF40AMiSaGQdkI725tXHHgNnLLlbR8afu3/WY4jNPrMbwhXmY4TTeIwMhcYig5HMlHV0fHlUiQWvgMBXT9K/09A31Ld+/HcXjmDoSIBXCuLlcPkdAXUP9anz5wWfLwZtYr2FrBNgqWUANrkiZNXKTac1DArAmMWQNr7iZr0naac/brM7f/YMqwaN6U6lFp0paNy057Tvi67lyFFISpNd2Z7kx3pru9644kNG3wpjvTnenOdLcD3V2XDw==', 'Position.z': '7ZQ7T8NAEIT/y9V2dPu6V08NEikgiCJCKSyUBCWmivzf8cV3UaqFgvIaS5b8aXZnx3MxT8fzMA7Hg0lvF7Me9rvzuN1/mWTQIvY29BjWwIk4SVx5x86D35jOPBzG07A7m3QxkB/P43b8nl/N42F92n58zp+8mASWsTOvJqGI7czGJAaxU2dIYUj8wiBQYZhnhhUGgxQmSmX8zIimg1wYgcrAzDiFkaIDkaoOZh2wmpB3ixBEuDcBNOcgFhfoBnHIkGaDs76MZ2Mdz2VI3ancCGIIdTzJUFCPBMU8rIfFfCTUjBCq40WuO2GGUHWvGAEh3luOaoY8V8trhuSqpLlHVENE1QjORpB2Jw4l4dZhgSgrsaKEMV6ZPrBb3KNwVWKvZdyHkgh724lyxrWQgyt3YqzuSb6T09xjobJTvBmRY+SV8bynBeqBrSs/h/M0TVP3e7NEZsfcmqU1S2uW1iz/1CySrF3l/X2U1iytWVqztGb5a7O8Tz8=', 'ExtrapolatedClock': {'Utc': '2022-08-28T21:43:42.3845884Z', 'Remaining': '00:00:00', 'Extrapolating': False}, 'TopThree': {'Withheld': False, 'Lines': [{'Position': '1', 'ShowPosition': True, 'RacingNumber': '1', 'Tla': 'VER', 'BroadcastName': 'M VERSTAPPEN', 'FullName': 'Max VERSTAPPEN', 'Team': 'Red Bull Racing', 'TeamColour': '1e5bc6', 'LapTime': '1:53.149', 'LapState': 64, 'DiffToAhead': 'LAP 44', 'DiffToLeader': 'LAP 44', 'OverallFastest': False, 'PersonalFastest': False}, {'Position': '2', 'ShowPosition': True, 'RacingNumber': '11', 'Tla': 'PER', 'BroadcastName': 'S PEREZ', 'FullName': 'Sergio PEREZ', 'Team': 'Red Bull Racing', 'TeamColour': '1e5bc6', 'LapTime': '1:52.843', 'LapState': 64, 'DiffToAhead': '+17.841', 'DiffToLeader': '+17.841', 'OverallFastest': False, 'PersonalFastest': False}, (... data excluded)
```

**Reply by theOehrly** [reply]
@oscr sorry for the slow response, I was really busy this week.

This looks fantastic. You can obviously tell that I wasn't really motivated to tackle this problem 😅 This certainly is an improvement if we get full functionality with all versions of Python. The question is, whether you'll try a pull request on the project. Maybe the dev is still active enough to merge it and release a new version. Although, I see that they currently list compatibility down to Python 3.4. If this breaks, it won't be quite that easy to just make a new release.
The other option is to just publish a new module from your fork, to skip dealing with all that backwards compatibility. What were your plans for this?

**Reply by oscr** [reply]
@theOehrly No problem. We all have lives outside GitHub :smile: 

I've asked maintainer if there could be a new release based on the hack I made: https://github.com/slazarov/python-signalr-client/issues/22

Let's see what/if he answers. If a new release is made then we can simply just bump the dependency.

**Reply by theOehrly** [reply]
@oscr great, let's see if there's a response.

**Reply by oscr** [reply]
@theOehrly The maintainer answered really quickly which was great, but the library is abandoned.

I looked at some of the other libraries but nothing seemed like an obvious alternative. I therefore suggest keeping library but making it a submodule for Fast-F1. Making minimal changes to get it Python 3.10 ready and stripping some extra stuff that isn't needed. It would require minimal changes to Fast-F1 itself. This allows changes if needed to the module but doesn't require them. If in the future signalr becomes huge and a great library shows up we can just delete the module and use the library itself.

**Reply by theOehrly** [reply]
@oscr I read the response. It's great that they answered this fast. 

If I understand you correct, you'd favour shipping the code directly with FastF1 instead of as a separate dependency? The only advantages that I see with this is that it reduces the effort required to do this. On the other hand though, a SignalR client isn't really a main concern of FastF1 and in terms of modularity, it should probably be kept separate. I'm somewhat undecided currently on how to best proceed with this.

**Reply by oscr** [reply]
@theOehrly I can fully understand why you wouldn't want it. I'm not saying it's a great choice, but in my opinion it's the least bad choice. Even if it's not a core concern for FF1 it's a core part in making it work. At least bringing it in makes it simple to manage and change if needed. This is not that uncommon. I recently cleaned out a 3rd party library from cluster-api. The alternative is otherwise that "someone" would need to publish and maintain a fork or implement another signalr library if one exists. I personally think that sounds worse :grin:

**Reply by theOehrly** [reply]
@oscr well, the "publish" part is not really the problem. The "maintain" part is. Questions is, how much maintenance is necessary. But it doesn't really make too much sense to publish it if I don't plan on maintaining it. Also, there doesn't seem to be much interest in SignalR. So it's likely mostly wasted effort to publish it.
The code is licensed under MIT as well, so that's fine. I guess you are right, and the best way to go forward is to just pull the code into FastF1.

**Reply by oscr** [reply]
@theOehrly Excellent. I think this will be great for beginners given Python 3.10 is the default version now being installed. I will make a draft these next few weeks.

**Reply by theOehrly** [reply]
> @theOehrly Excellent. I think this will be great for beginners given Python 3.10 is the default version now being installed. I will make a draft these next few weeks.

Awesome, looking forward to that 👍

**Reply by borolepratik** [reply]
@oscr checking in to see if there are any updates.

**Reply by oscr** [reply]
@borolepratik The work is completed and in master: https://github.com/theOehrly/Fast-F1/pull/254 It is part of the upcoming 3.0.0 release: https://github.com/theOehrly/Fast-F1/milestone/2?closed=1

**Comment by foxyskippy** [comment]
I just upgrade the FastF1 package to v2.3.0 but it didn't work. Can I try something else?

**Reply by oscr** [reply]
@foxyskippy A possible fix is being discussed but there is nothing merged yet. Depending on your experience level there are ways of running different Python versions. But maybe pyenv would work for you https://github.com/pyenv/pyenv ? There are of course other ways such as Docker or asdf.
